crates/runtime/src/backtrace.rs

```diff
@@ -0,0 +1,251 @@
+            assert_eq!(first_wasm_sp % 8, 0);
+            assert_eq!(first_wasm_sp % 16, 8);
+        } else if cfg!(target_arch = "aarch64") {
+            assert_eq!(first_wasm_sp % 16, 0);
+        }
+
+        loop {
+            assert_eq!(fp % 16, 0, "stack should always be aligned to 16");
+
+            log::trace!("--- Tracing through one Wasm frame ---");
+            log::trace!("pc = 0x{:016x}", pc);
+            log::trace!("fp = 0x{:016x}", fp);
+
+            f(Frame { pc, fp })?;
+
+            // If our FP has reached the SP upon entry to Wasm from the
+            // host, then we've successfully walked all the Wasm frames,
+            // and have now reached a host frame. We're done iterating
+            // through this contiguous sequence of Wasm frames.
+            if Self::reached_entry_sp(fp, first_wasm_sp) {
+                return ControlFlow::Continue(());
+            }
+
+            // If we didn't return above, then we know we are still in a
+            // Wasm frame, and since Cranelift maintains frame pointers,
+            // we know that the FP isn't an arbitrary value and it is
+            // safe to dereference it to read the next PC/FP.
+
+            // The calling convention always pushes the return pointer
+            // (aka the PC of the next older frame) just before this
+            // frame.
+            pc = Self::get_next_older_pc_from_fp(fp);
+
+            // And the current frame pointer points to the next older
+            // frame pointer. Because the stack grows down, the older FP
+            // must be greater than the current FP.
+            let next_older_fp = Self::get_next_older_fp_from_fp(fp);
+            assert!(next_older_fp > fp, "{next_older_fp:#x} > {fp:#x}");
+            fp = next_older_fp;
+        }
+    }
+
+    unsafe fn get_next_older_pc_from_fp(fp: usize) -> usize {
+        cfg_if! {
+            if #[cfg(target_arch = "x86_64")] {
+                *(fp as *mut usize).offset(1)
+            } else if #[cfg(target_arch = "aarch64")] {
+                *(fp as *mut usize).offset(1)
+            } else {
+                compile_error!("platform not supported")
+            }
+        }
+    }
+
+    unsafe fn get_next_older_fp_from_fp(fp: usize) -> usize {
+        cfg_if! {
+            if #[cfg(target_arch = "x86_64")] {
+                *(fp as *mut usize)
+            } else if #[cfg(target_arch = "aarch64")] {
+                *(fp as *mut usize)
+            } else {
+                compile_error!("platform not supported")
+            }
+        }
+    }
+
+    unsafe fn reached_entry_sp(fp: usize, first_wasm_sp: usize) -> bool {
+        cfg_if! {
+            if #[cfg(target_arch = "x86_64")] {
+                fp == first_wasm_sp - 8
+            } else if #[cfg(target_arch = "aarch64")] {
+                fp == first_wasm_sp - 16
+            } else {
+                compile_error!("platform not supported")
+            }
+        }
+    }
```

Comment:
I mentioned this before, but personally I think it would be best to cut down on the `cfg_if!` here. I'd rather not have a duplicated set within each function. Could this instead move to the architecture-specific modules in the `wasmtime-runtime` crate?

Reply:
Split everything out into modules, but haven't added non-x64 docs yet (pinged folks for help with that).